model/riscv_insts_vext_utils.sail

```diff
@@ -0,0 +1,703 @@
+}
+
+/* Floating point canonical NaN for 16-bit, 32-bit, 64-bit and 128-bit types */
+val canonical_NaN : forall 'm, 'm in {16, 32, 64, 128}. int('m) -> bits('m)
+function canonical_NaN('m) = {
+  match 'm {
+    16   => 0x_7e00,
+    32   => 0x_7fc0_0000,
+    64   => 0x_7ff8_0000_0000_0000,
+    128  => 0x_7fff_8000_0000_0000_0000_0000_0000_0000
+  }
+}
+
+/* Floating point NaN boxing / unboxing that support 16-bit to 128-bit types */
+val nan_box_new : forall 'm, ('m in {16, 32, 64, 128} & 'm <= flen). bits('m) -> flenbits
+function nan_box_new   unboxed = {
+  if sizeof(flen) == 'm
+  then unboxed
+  else sail_ones(sizeof(flen) - 'm) @ unboxed 
+}
+
+val nan_unbox_new : forall 'm, ('m in {16, 32, 64, 128} & 'm <= flen). (flenbits, int('m)) -> bits('m)
+function nan_unbox_new(regval, 'm) = {
+  if sizeof(flen) == 'm
+  then regval
+  else if slice(regval, 'm, sizeof(flen) - 'm) == ones()
+  then slice(regval, 0, 'm)
+  else canonical_NaN('m)
+}
+
+/* Check if the floating point number is a signaling NaN */
+val      f_is_SNaN : forall 'm, ('m in {16, 32, 64, 128} & 'm <= flen). bits('m) -> bool
+function f_is_SNaN   xf = {
+  match 'm {
+    16   => is_ones(xf[14..10]) & is_zeros(xf[9..9]) & ~(is_zeros(xf[8..0])),
+    32   => is_ones(xf[30..23]) & is_zeros(xf[22..22]) & ~(is_zeros(xf[21..0])),
+    64   => is_ones(xf[62..52]) & is_zeros(xf[51..51]) & ~(is_zeros(xf[50..0])),
+    128  => is_ones(xf[126..112]) & is_zeros(xf[111..111]) & ~(is_zeros(xf[110..0]))
+  }
+}
+
+/* Either QNaN or SNan */
+val      f_is_NaN : forall 'm, ('m in {16, 32, 64, 128} & 'm <= flen). bits('m) -> bool
+function f_is_NaN   xf = {
+  match 'm {
+    16   => is_ones(xf[14..10]) & ~(is_zeros(xf[9..0])),
+    32   => is_ones(xf[30..23]) & ~(is_zeros(xf[22..0])),
+    64   => is_ones(xf[62..52]) & ~(is_zeros(xf[51..0])),
+    128  => is_ones(xf[126..112]) & ~(is_zeros(xf[111..0]))
+  }
+}
+
+val      f_is_neg_zero : forall 'm, ('m in {16, 32, 64, 128} & 'm <= flen). bits('m) -> bool
+function f_is_neg_zero   xf = {
+  match 'm {
+    16   => is_ones(xf[15..15]) & is_zeros(xf[14..0]),
+    32   => is_ones(xf[31..31]) & is_zeros(xf[30..0]),
+    64   => is_ones(xf[63..63]) & is_zeros(xf[62..0]),
+    128  => is_ones(xf[127..127]) & is_zeros(xf[126..0])
+  }
+}
+
+val      f_is_pos_zero : forall 'm, ('m in {16, 32, 64, 128} & 'm <= flen). bits('m) -> bool
+function f_is_pos_zero   xf = {
+  match 'm {
+    16   => is_zeros(xf),
+    32   => is_zeros(xf),
+    64   => is_zeros(xf),
+    128  => is_zeros(xf)
+  }
+}
+
+/* Scalar register shaping for floating point operations */
+val get_scalar_fp : forall 'n, 'n in {16, 32, 64, 128}. (regidx, int('n)) -> bits('n) effect {escape, rreg}
+function get_scalar_fp(rs1, vsew_bits) = {
+  if sizeof(xlen) >= vsew_bits then {
+    /* Least significant SEW bits */
+    nan_unbox_new(F(rs1), vsew_bits)
+  }
+  else {
+    canonical_NaN(vsew_bits)
+  }
+}
+
+val log2 : forall 'n, 8 <= 'n <= 1024. int('n) -> int
+function log2(n) = {
```

Comment:
This feels like something more for prelude or similar?

Reply:
This function is now moved to prelude.sail.